File: src/tools/plans.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import GraphClient from '../graph-client.js';

export function registerPlanTools(server: McpServer, graphClient: GraphClient): void {
  server.tool(
    'create-plan',
    'Create a new Planner plan. The owner must be a Microsoft 365 Group ID.',
    {
      title: z.string().describe('Plan title'),
      owner: z.string().describe('Group ID that owns the plan (use list-groups to find)'),
    },
    { title: 'create-plan', destructiveHint: true, openWorldHint: true },
    async ({ title, owner }) => {
      try {
        const result = await graphClient.post('/planner/plans', { owner, title });
        return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
      } catch (error) {
        return { content: [{ type: 'text', text: JSON.stringify({ error: (error as Error).message }) }], isError: true };
      }
    }
  );

  server.tool(
    'update-plan',
    'Update a Planner plan (title, category descriptions). ETag is auto-fetched if not provided.',
    {
      'plan-id': z.string().describe('Plan ID'),
      title: z.string().optional().describe('New plan title'),
      categoryDescriptions: z.record(z.string()).optional().describe('Category label descriptions, e.g. {"category1": "Urgent", "category2": "Bug"}'),
      etag: z.string().optional().describe('ETag for optimistic concurrency (auto-fetched if omitted)'),
    },
    { title: 'update-plan', destructiveHint: true, openWorldHint: true },
    async (params) => {
      try {
        const planId = params['plan-id'];
        const etag = params.etag || await graphClient.getEtag(`/planner/plans/${planId}`);
        const body: Record<string, unknown> = {};
        if (params.title) body.title = params.title;
        if (params.categoryDescriptions) body.categoryDescriptions = params.categoryDescriptions;
        const result = await graphClient.patch(`/planner/plans/${planId}`, body, etag);
        return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
      } catch (error) {
        return { content: [{ type: 'text', text: JSON.stringify({ error: (error as Error).message }) }], isError: true };
      }
    }
  );

  server.tool(
    'delete-plan',
    'Delete a Planner plan. ETag is auto-fetched if not provided.',
    {
      'plan-id': z.string().describe('Plan ID'),
      etag: z.string().optional().describe('ETag for optimistic concurrency (auto-fetched if omitted)'),
    },
    { title: 'delete-plan', destructiveHint: true, openWorldHint: true },
    async (params) => {
      try {
        const planId = params['plan-id'];
        const etag = params.etag || await graphClient.getEtag(`/planner/plans/${planId}`);
        await graphClient.delete(`/planner/plans/${planId}`, etag);
        return { content: [{ type: 'text', text: JSON.stringify({ success: true, message: 'Plan deleted' }) }] };
      } catch (error) {
        return { content: [{ type: 'text', text: JSON.stringify({ error: (error as Error).message }) }], isError: true };
      }
    }
  );
}

File: go/internal/tools/plans.go
package tools

import (
	"context"
	"fmt"

	"github.com/TestaVivaDK/plannner-connector/internal/graph"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func registerPlanTools(s *server.MCPServer, gc *graph.Client) {
	// create-plan
	createPlan := mcp.NewTool("create-plan",
		mcp.WithDescription("Create a new Planner plan. The owner must be a Microsoft 365 Group ID."),
		mcp.WithString("title", mcp.Required(), mcp.Description("Plan title")),
		mcp.WithString("owner", mcp.Required(), mcp.Description("Group ID that owns the plan (use list-groups to find)")),
		mcp.WithDestructiveHintAnnotation(true),
	)
	s.AddTool(createPlan, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		title, err := req.RequireString("title")
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}
		owner, err := req.RequireString("owner")
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}
		raw, err := gc.Post("/planner/plans", map[string]any{"title": title, "owner": owner})
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf(`{"error":"%s"}`, err)), nil
		}
		return mcp.NewToolResultText(formatJSON(raw)), nil
	})

	// update-plan
	updatePlan := mcp.NewTool("update-plan",
		mcp.WithDescription("Update a Planner plan (title, category descriptions). ETag is auto-fetched if not provided."),
		mcp.WithString("plan-id", mcp.Required(), mcp.Description("Plan ID")),
		mcp.WithString("title", mcp.Description("New plan title")),
		mcp.WithObject("categoryDescriptions", mcp.Description(`Category label descriptions, e.g. {"category1": "Urgent", "category2": "Bug"}`)),
		mcp.WithString("etag", mcp.Description("ETag for optimistic concurrency (auto-fetched if omitted)")),
		mcp.WithDestructiveHintAnnotation(true),
	)
	s.AddTool(updatePlan, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		planID, err := req.RequireString("plan-id")
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}
		etag := req.GetString("etag", "")
		if etag == "" {
			etag, err = gc.GetEtag("/planner/plans/" + planID)
			if err != nil {
				return mcp.NewToolResultError(fmt.Sprintf(`{"error":"%s"}`, err)), nil
			}
		}
		body := make(map[string]any)
		if v := req.GetString("title", ""); v != "" {
			body["title"] = v
		}
		args := req.GetArguments()
		if v, ok := args["categoryDescriptions"]; ok {
			body["categoryDescriptions"] = v
		}
		raw, err := gc.Patch("/planner/plans/"+planID, body, etag)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf(`{"error":"%s"}`, err)), nil
		}
		return mcp.NewToolResultText(formatJSON(raw)), nil
	})

	// delete-plan
	deletePlan := mcp.NewTool("delete-plan",
		mcp.WithDescription("Delete a Planner plan. ETag is auto-fetched if not provided."),
		mcp.WithString("plan-id", mcp.Required(), mcp.Description("Plan ID")),
		mcp.WithString("etag", mcp.Description("ETag for optimistic concurrency (auto-fetched if omitted)")),
		mcp.WithDestructiveHintAnnotation(true),
	)
	s.AddTool(deletePlan, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		planID, err := req.RequireString("plan-id")
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}
		etag := req.GetString("etag", "")
		if etag == "" {
			etag, err = gc.GetEtag("/planner/plans/" + planID)
			if err != nil {
				return mcp.NewToolResultError(fmt.Sprintf(`{"error":"%s"}`, err)), nil
			}
		}
		_, err = gc.Delete("/planner/plans/"+planID, etag)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf(`{"error":"%s"}`, err)), nil
		}
		return jsonResult(map[string]any{"success": true, "message": "Plan deleted"}), nil
	})
}
